nuxeo-runtime/nuxeo-runtime-cluster/src/main/java/org/nuxeo/runtime/cluster/ClusterServiceImpl.java

```diff
@@ -117,4 +123,148 @@ public void setNodeId(String nodeId) {
+                        TransactionHelper.startTransaction();
+                    } finally {
+                        if (completedAbruptly) {
+                            // mark rollback-only if there was an exception
+                            TransactionHelper.setTransactionRollbackOnly();
+                        }
+                    }
+                }
+            }
+        }
+
+        /**
+         * Runs a {@link Runnable} atomically, in a cluster-wide critical section.
+         */
+        protected void runAtomicallyInternal(String key, Runnable runnable) {
+            String lockInfo = tryLock(key);
+            if (lockInfo != null) {
+                try {
+                    runnable.run();
+                } finally {
+                    unLock(key, lockInfo);
+                }
+            } else {
+                throw new RuntimeServiceException("Failed to acquire lock '" + key + "' after " + duration.toSeconds()
+                        + "s, owner: " + getLock(key));
+            }
+        }
+
+        // try to acquire the lock and fail if it takes too long
+        protected String tryLock(String key) {
+            log.debug("Trying to lock '{}'", key);
+            long deadline = System.nanoTime() + duration.toNanos();
+            long ttl = duration.multipliedBy(TTL_MULTIPLIER).toSeconds();
+            do {
+                // try to acquire the lock
+                String lockInfo = "node=" + nodeId + " time=" + Instant.now();
+                if (kvStore.compareAndSet(key, null, lockInfo, ttl)) {
+                    // lock acquired
+                    log.debug("Lock '{}' acquired after {}ms", () -> key,
+                            () -> (System.nanoTime() - (deadline - duration.toNanos())) / 1_000_000);
+                    return lockInfo;
+                }
+                // wait a bit before retrying
+                log.debug("  Sleeping on busy lock '{}' for {}ms", () -> key, () -> pollDelay.toMillis());
+                try {
+                    Thread.sleep(pollDelay.toMillis());
+                } catch (InterruptedException e) {
+                    Thread.currentThread().interrupt();
+                    throw new RuntimeServiceException(e);
+                }
+            } while (System.nanoTime() < deadline);
+            log.debug("Failed to acquire lock '{}' after {}s", () -> key, () -> duration.toSeconds());
+            return null;
+        }
+
+        protected void unLock(String key, String lockInfo) {
+            log.debug("Unlocking '{}'", key);
+            if (kvStore.compareAndSet(key, lockInfo, null)) {
+                return;
+            }
+            // couldn't remove the lock, it expired an may have been reacquired
+            String current = kvStore.getString(key);
+            if (current == null) {
+                // lock expired but was not reacquired
+                log.warn("Unlocking '{}' but the lock had already expired; "
```

Comment:
`Unlocking '{}' but the lock had already expired; ` -> `Unlocking '{}' but the lock had already expired, ` a `,` instead of `;`

Reply:
The `;` is on purpose for better grammar.